[c10d FR analyzer] Output a meaningful debug report for users

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134528

- This PR generates a more useful output log for users: P1552399180.
- It also fixes the logic when we check the all-gather size mismatch.
- Add dtype check for collective input/output
- We store more context information for error match_state so that we can report them in the file.
- Disable the size match for alltoall because we don't log the size for all inputs/outputs.
- Correct some types for func args specification.


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d FR analyzer] Output a meaningful debug report for users

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d FR analyzer] Output a meaningful debug report for users"


This PR generates a more useful output log for users: P1549841172.
It also fixes the logic when we check the all-gather size mismatch.
We store more context information for error match_state so that we can report them in the file.


[ghstack-poisoned]

fduwjj committed:
Update on "[c10d FR analyzer] Output a meaningful debug report for users"


- This PR generates a more useful output log for users: P1552399180.
- It also fixes the logic when we check the all-gather size mismatch.
- Add dtype check for collective input/output
- We store more context information for error match_state so that we can report them in the file.
- Disable the size match for alltoall because we don't log the size for all inputs/outputs.
- Correct some types for func args specification.


[ghstack-poisoned]

c-p-i-o commented on tools/flight_recorder/components/builder.py:
```diff
@@ -284,11 +300,17 @@ def build_collectives(
                     )
+                    # We don't log the input/output sizes for alltoall so we don't consider the size mismatch as an error for now.
+                    check_result = True
                     if not check_result:
```
This `if` feels backwards?
I would have expected you to do something like:
```
check_result = False
if check_result:
```

fduwjj replied:
maybe I can rename it to `fail_check`

c-p-i-o commented on tools/flight_recorder/components/builder.py:
```diff
@@ -284,11 +300,17 @@ def build_collectives(
+                            f"for group {pg_name}:{desc} collective {profiling_name} ",
+                            f"input_numel {input_numel} output_numel {output_numel} ",
+                            f"{input_sizes} {output_sizes} {len(expected_ranks)} {collective_state} ",
+                            f"\nCollective stack traces: \n{colletive_frames}",
```
Typo here = unless it needs to be fixed elsewhere?
```suggestion
                            f"\nCollective stack traces: \n{collective_frames}",
```

c-p-i-o commented on tools/flight_recorder/components/builder.py:
```diff
@@ -306,11 +328,16 @@ def build_collectives(
+                    f"Collective {record_id} errors for group {pg_name}:{desc} collective {profiling_name} ",
+                    f"{input_sizes} {output_sizes} {len(expected_ranks)} {collective_state} ",
+                    f"\nFound errors: {error_msg}\n",
+                    f"\nCollective stack traces: \n{colletive_frames} ",
```
Same as above. typo `colletive` -> `collective`

c-p-i-o commented on tools/flight_recorder/components/utils.py:
```diff
@@ -216,16 +230,46 @@ def check_version(versions: Dict[str, Any]) -> None:
+        # find the largest starting point. For example, if the buffer has the following entries:
+        # Rank 0: [0, 1, 2, 3, 4, 5, 6]
+        # Rank 1: [7, 1, 2, 3, 4, 5, 6]
+        # Rank 2: [7, 8, 2, 3, 4, 5, 6]
```
One thing to note is the expectation is entry `0` is the lowest record id and entry `N` is the largest record id. FR entries are supposed to be sorted when dumped.
So, we should not see 7, 8, 1, 2, 3 .

fduwjj replied:
Oh right, I missed that part, inside `dump_entries` we already have sorting already, then we don't need to re-sort it again.

c-p-i-o reviewed (approved):
Minor comments - but LGTM!

fduwjj committed:
Update on "[c10d FR analyzer] Output a meaningful debug report for users"


- This PR generates a more useful output log for users: P1552399180.
- It also fixes the logic when we check the all-gather size mismatch.
- Add dtype check for collective input/output
- We store more context information for error match_state so that we can report them in the file.
- Disable the size match for alltoall because we don't log the size for all inputs/outputs.
- Correct some types for func args specification.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
